src/main/java/org/folio/auth/authtokenmodule/storage/RefreshTokenStore.java

```diff
@@ -0,0 +1,216 @@
+
+    log.info("Creating {} tables", RefreshTokenStore.class.getName());
+
+    return pool.query(createIndexExpiresAt).execute()
+      .compose(x -> pool.query(createIndexUserId).execute())
+      .mapEmpty();
+  }
+
+  /**
+   * Save the token to the token store. This should be done anytime a new refresh token
+   * is issued.
+   * @param refreshToken The refresh token to store.
+   * @return A failed future should the save operation fail. Otherwise a succeeded future
+   * is returned.
+   */
+  public Future<Void> saveToken(RefreshToken refreshToken) {
+    UUID id = refreshToken.getId();
+    UUID userId = refreshToken.getUserId();
+    long expiresAt = refreshToken.getExpiresAt();
+    String tenant = refreshToken.getTenant();
+    boolean isRevoked = false;
+    boolean isRedeemed = false;
+
+    log.info("Inserting token id {} into {} token store: {}", id, REFRESH_TOKEN_SUFFIX, expiresAt);
+
+    String insert = "INSERT INTO " + tableName(tenant, REFRESH_TOKEN_SUFFIX) +
+        "(id, user_id, is_revoked, is_redeemed, expires_at) VALUES ($1, $2, $3, $4, $5)";
+    var values = Tuple.of(id, userId, isRevoked, isRedeemed, expiresAt);
+
+    return pool.preparedQuery(insert).execute(values).mapEmpty();
+  }
+
+  /**
+   * Check that the token has not been revoked. This will return a failed future if
+   * the token has been revoked, otherwise it will return a succeeded future.
+   *
+   * This is what this method does:
+   * 1. Get isRedeemed and isRevoked for token. If token not found treat as revoked.
+   * 2. If it has been revoked, fail the future. We need to check this here since
+   *    other requests could have revoked the token anytime. We can't let these through.
+   * 3. If isRedeemed is false, call setTokenRedeemed and return a success. It is not
+   *    revoked or redeemed. This is the desired state. RTs can be used only once.
+   *    Any subsequent uses must be considered as a leaked (compromised) token.
+   * 4. If isRedeemed is true call revokeAllTokensForUser and return a failed future.
+   *    isRedeemed true means "someone tried to use it a second time". This is the
+   *    second use of the token and it means the token has been leaked. We
+   *    consider the user's account compromised and revoke all their refresh tokens.
+   *    They will need to login again to get a new RefreshToken that isn't revoked
+   *    or redeemed.
+   *
+   * @param refreshToken The RefreshToken to check.
+   * @return A failed future if the token has been revoked. Otherwise a succeeded future
+   * is returned.
+   */
+  public Future<Void> checkTokenNotRevoked(RefreshToken refreshToken) {
+    UUID tokenId = refreshToken.getId();
+    UUID userId = refreshToken.getUserId();
+    String table = tableName(tenant, REFRESH_TOKEN_SUFFIX);
+
+    log.info("Checking token redeemed of token id {} for tenant {}", tokenId, tenant);
+
+    String select = "SELECT is_redeemed, is_revoked FROM " + table + "WHERE id=$1";
```

Comment:
Sending two request with the same refresh token in parallel to mod-authtoken may successfully create two new refresh tokens because there is a race condition in the time interval between the SELECT and the UPDATE statement.

To avoid the race condition they can be combined into an atomic operation:

```
UPDATE table SET is_redeemed=TRUE WHERE id=$1 AND is_redeemed=FALSE and is_revoked=FALSE
```

The returned rowCount is either 1 or 0.
1 indicates success.
0 indicates that the refresh token doesn't exists, or that is_redeemed or is_revoked is true. This is a failure. Simply SELECT the record and take the needed actions.

Reply:
Can you add a check for expires_at to reject the token if it is expired?

Reply:
This is a great idea and also a nice simplification since there's no longer a need to select anything either before or after.

> Can you add a check for expires_at to reject the token if it is expired?

Added. I opted to just check the provided rather than the stored token since the provided token shouldn't have been tampered with. See what you think.